hail/src/main/scala/is/hail/expr/types/physical/PCanonicalArray.scala

```diff
@@ -0,0 +1,368 @@
+  }
+
+  def setAllMissingBits(region: Region, aoff: Long, length: Int) {
+    if (elementType.required)
+      return
+    writeMissingness(region, aoff, length, -1)
+  }
+
+  def clearMissingBits(region: Region, aoff: Long, length: Int) {
+    if (elementType.required)
+      return
+    writeMissingness(region, aoff, length, 0)
+  }
+
+  def initialize(region: Region, aoff: Long, length: Int, setMissing: Boolean = false) {
+    Region.storeInt(aoff, length)
+    if (setMissing)
+      setAllMissingBits(region, aoff, length)
+    else
+      clearMissingBits(region, aoff, length)
+  }
+
+  def stagedInitialize(aoff: Code[Long], length: Code[Int], setMissing: Boolean = false): Code[Unit] = {
+    if (elementType.required)
+      Region.storeInt(aoff, length)
+    else
+      Code(
+        Region.storeInt(aoff, length),
+        Region.setMemory(aoff + const(lengthHeaderBytes), nMissingBytes(length).toL, const(if (setMissing) (-1).toByte else 0.toByte)))
+  }
+
+  def zeroes(region: Region, length: Int): Long = {
+    require(elementType.isNumeric)
+    val aoff = allocate(region, length)
+    initialize(region, aoff, length)
+    Region.setMemory(aoff + elementsOffset(length), length * elementByteSize, 0.toByte)
+    aoff
+  }
+
+  def zeroes(mb: MethodBuilder, region: Code[Region], length: Code[Int]): Code[Long] = {
+    require(elementType.isNumeric)
+    val aoff = mb.newLocal[Long]
+    Code(
+      aoff := allocate(region, length),
+      stagedInitialize(aoff, length),
+      Region.setMemory(aoff + elementsOffset(length), length.toL * elementByteSize, 0.toByte),
+      aoff)
+  }
+
+  def anyMissing(mb: MethodBuilder, aoff: Code[Long]): Code[Boolean] =
+    if (elementType.required)
+      false
+    else {
+      val n = mb.newLocal[Long]
+      JoinPoint.CallCC[Code[Boolean]] { (jb, ret) =>
+        val loop = jb.joinPoint[Code[Long]](mb)
+        loop.define { ptr =>
+          (ptr < n).mux(
+            Region.loadInt(ptr).cne(0).mux(
+              ret(true),
+              loop(ptr + 4L)),
+            (Region.loadByte(ptr) >>>
+              (const(32) - (loadLength(aoff) | 31))).cne(0).mux(
+              ret(true),
+              ret(false)))
+        }
+        Code(
+          n := aoff + ((loadLength(aoff) >>> 5) * 4 + 4).toL,
+          loop(aoff + 4L))
+      }
+    }
+
+  def forEach(mb: MethodBuilder, region: Code[Region], aoff: Code[Long], body: Code[Long] => Code[Unit]): Code[Unit] = {
+    val i = mb.newLocal[Int]
+    val n = mb.newLocal[Int]
+    Code(
+      n := loadLength(aoff),
+      i := 0,
+      Code.whileLoop(i < n,
+        isElementDefined(aoff, i).mux(
+          body(loadElement(region, aoff, n, i)),
+          Code._empty
+        )))
+  }
+
+  override def unsafeOrdering(): UnsafeOrdering =
+    unsafeOrdering(this)
+
+  override def unsafeOrdering(rightType: PType): UnsafeOrdering = {
+    val right = rightType.asInstanceOf[PContainer]
```

Comment:
In order to make this work in PSet and PDict, I needed to remove the `require(this.isOfType(rightType))`. I'm also not sure this requirement should be in place, the cast below guarantees that this occurs in a PContainer inheritor, which means both conform to at least the interface needed (since we only need loadLength method).

Otherwise it fails in UnsafeSuite when we try to order a set and an array:

```
Unsafe ordering set<tuple(struct{})>, array<tuple(struct{})>
```
(imagine this debug message is added to the top of the unsafeOrdering function in PCanonicalArray).